rust/private/unpretty.bzl

```diff
@@ -0,0 +1,380 @@
+    inputs = [str(Label("//rust/settings:unpretty"))],
+    outputs = [
+        str(Label("//rust/settings:unpretty")),
+        str(Label("//rust/toolchain/channel")),
+    ],
+)
+
+def _get_unpretty_ready_crate_info(target, aspect_ctx):
+    """Check that a target is suitable for expansion and extract the `CrateInfo` provider from it.
+
+    Args:
+        target (Target): The target the aspect is running on.
+        aspect_ctx (ctx, optional): The aspect's context object.
+
+    Returns:
+        CrateInfo, optional: A `CrateInfo` provider if rust unpretty should be run or `None`.
+    """
+
+    # Ignore external targets
+    if target.label.workspace_root.startswith("external"):
+        return None
+
+    # Targets with specific tags will not be formatted
+    if aspect_ctx:
+        ignore_tags = [
+            "nounpretty",
+            "no-unpretty",
+            "no_unpretty",
+        ]
+
+        for tag in ignore_tags:
+            if tag in aspect_ctx.rule.attr.tags:
+                return None
+
+    # Obviously ignore any targets that don't contain `CrateInfo`
+    if rust_common.crate_info not in target:
+        return None
+
+    return target[rust_common.crate_info]
+
+def _rust_unpretty_aspect_impl(target, ctx):
+    crate_info = _get_unpretty_ready_crate_info(target, ctx)
+    if not crate_info:
+        return []
+
+    toolchain = find_toolchain(ctx)
+    cc_toolchain, feature_configuration = find_cc_toolchain(ctx)
+
+    dep_info, build_info, linkstamps = collect_deps(
+        deps = crate_info.deps,
+        proc_macro_deps = crate_info.proc_macro_deps,
+        aliases = crate_info.aliases,
+        # Rust expand doesn't need to invoke transitive linking, therefore doesn't need linkstamps.
+        are_linkstamps_supported = False,
+    )
+
+    compile_inputs, out_dir, build_env_files, build_flags_files, linkstamp_outs, ambiguous_libs = collect_inputs(
+        ctx,
+        ctx.rule.file,
+        ctx.rule.files,
+        linkstamps,
+        toolchain,
+        cc_toolchain,
+        feature_configuration,
+        crate_info,
+        dep_info,
+        build_info,
+    )
+
+    output_groups = {}
+    outputs = []
+
+    for mode in ctx.attr._unpretty_modes[RustUnprettyInfo].modes.to_list():
+        pretty_mode = mode.replace("-", "_")
+        mnemonic = "RustUnpretty{}".format("".join([
+            o.title()
+            for m in pretty_mode.split(",")
+            for o in m.split("_")
+        ]))
+
+        unpretty_out = ctx.actions.declare_file(
+            "{}.unpretty.{}.rs".format(ctx.label.name, pretty_mode.replace(",", ".")),
+            sibling = crate_info.output,
+        )
+
+        output_groups.update({"rust_unpretty_{}".format(pretty_mode.replace(",", "_")): depset([unpretty_out])})
+        outputs.append(unpretty_out)
+
+        rust_flags = []
+        if crate_info.is_test:
+            if getattr(ctx.rule.attr, "use_libtest_harness", True):
```

Comment:
This feels like code that should exist somewhere shared, rather than being repeated here and in `rust.bzl` (and incorrectly repeated in `clippy.bzl` without the conditional)?

Reply:
Added a helper function. I'll update clippy in another PR